langfuse/openai.py

```diff
@@ -648,3 +602,97 @@ def _filter_image_data(messages: List[dict]):
                         del content[index]["image_url"]
 
     return output_messages
+
+
+class LangfuseResponseGeneratorSync:
+    def __init__(
+        self,
+        *,
+        resource,
+        response,
+        generation,
+        langfuse,
+        is_nested_trace,
+    ):
+        self.items = []
+
+        self.resource = resource
+        self.response = response
+        self.generation = generation
+        self.langfuse = langfuse
+        self.is_nested_trace = is_nested_trace
+
+    def __iter__(self):
+        try:
+            for i in self.response:
+                self.items.append(i)
+
+                yield i
+        finally:
+            self._finalize()
+
+    def __enter__(self):
+        return self.__iter__()
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        pass
+
+    def _finalize(self):
+        model, completion_start_time, completion = _extract_openai_response(
+            self.resource, self.items
+        )
+
+        # Avoiding the trace-update if trace-id is provided by user.
+        if not self.is_nested_trace:
+            self.langfuse.trace(id=self.generation.trace_id, output=completion)
+
+        _create_langfuse_update(
+            completion, self.generation, completion_start_time, model=model
+        )
+
+
+class LangfuseResponseGeneratorAsync:
```

Comment:
what about typing this as the openai sdk does it? We could subclass from `AsyncResponseContextManager`